pyswarms/utils/functions/grid_search.py

```diff
@@ -0,0 +1,102 @@
+# -*- coding: utf-8 -*-
+
+"""
+Hyperparameter grid search.
+
+Compares the relative performance of hyperparameter value combinations in
+reducing a specified objective function.
+
+For each hyperparameter, user can provide either a singe value or a list
+of possible values, and their cartesian product is taken to produce a grid
+of all possible combinations. These combinations are then tested to produce
+a list of objective function scores. The default of the optimize method
+returns the hyperparameters that yield the minimum score, yet maximum score
+can also be assessed.
+"""
+
+import operator as op
+import itertools
+import numpy as np
+
+class GridSearch():
+    """
+    Exhaustive search of optimal performance on selected objective function
+    over all combinations of specified hyperparameter values.
+
+    Parameters:
+        * c1 : float
+            cognitive parameter
+        * c2 : float
+            social parameter
+        * w : float
+            inertia parameter
+        * k : int
+            number of neighbors to be considered. Must be a
+            positive integer less than `n_particles`
+        * p: int {1,2}
+            the Minkowski p-norm to use. 1 is the
+            sum-of-absolute values (or L1 distance) while 2 is
+            the Euclidean (or L2) distance.
+    >>> options = {'c1': [1, 2, 3],
+                   'c2': [1, 2, 3],
+                   'w' : [2, 3, 5],
+                   'k' : [5, 10, 15],
+                   'p' : 1}
+    >>> g = GridSearch(LocalBestPSO, n_particles=40, dimensions=20,
+                       options=options, objective_func=sphere_func, iters=10)
+    >>> best_options = g.optimize()
+    >>> best_options['c1']
+    1
+    >>> best_options['c2']
+    1
+    """
+
+    def __init__(self, optimizer, n_particles, dimensions, options,
+                 objective_func, iters, bounds=None, velocity_clamp=None):
+        """Initializes with dict of key, parameter value pairs."""
+        self._optimizer = optimizer
+        self._n_particles = n_particles
+        self._dims = dimensions
+        self._options = options
+        self._bounds = bounds
+        self._vclamp = velocity_clamp
+        self._objective_func = objective_func
+        self._iters = iters
```

Comment:
I think it's fine to make these attributes "public" since its only the class that will access it for now. Just a matter of consistency with the other classes that we have. It's fine to remove the underscores. 😄

Reply:
I'm a big fan of consistency :+1:, and have removed the underscores.